### ISSUE TITLE
display messy code

### ISSUE THREAD

**Issue by git-jiadong** [kingToolbox/WindTerm#162]
![image](https://user-images.githubusercontent.com/30514201/129678416-f5586623-40dd-4ba9-9877-cf5c50fa9d2a.png)

**Comment by git-jiadong** [comment]
打印日志出现乱码，应该如何恢复，目前没有找到比较好的办法，只能重启会话

**Comment by git-jiadong** [comment]
![image](https://user-images.githubusercontent.com/30514201/129687374-37581356-aeef-4eae-b126-98921fd8fb2d.png)
使用其他终端操作如图不会出现乱码，但使用windterm会出现乱码

**Comment by kingToolbox** [comment]
Thank you. This is a known issue. But I haven't figured out how to deal with it. 

First of all, not only WindTerm, all clients/terminals usually cannot print binary data directly. In other words, `cat` a binary file in any terminal may cause garbled characters to be displayed. This is because the TERM control sequences were not designed to be compatible with the binary sequence. 

Second, some control sequences define how to switch character ses. 
- SI (\xe):
Switch to Standard Character Set (Ctrl-O is Shift In or LS0). This invokes the G0 character set (the default) as GL. VT200 and up implement LS0.
- SO (\xf)
Switch to Alternate Character Set (Ctrl-N is Shift Out or LS1).  This invokes the G1 character set as GL. VT200 and up implement LS1.

So when the received binary data happens to contain the control sequence to switch the character set, the terminal will switch the character set. Although this is correct, it is not what the user wants.

WindTerm set the G1 character set to [DEC supplemental graphics](https://vt100.net/docs/vt220-rm/table2-4.html) according to the standard, so after switching, the characters are displayed as table characters. I will check what is the default G1 character set defined by other terminals. 

In dealing with binary sequences, WindTerm is indeed a little less robust than other clients. Because unlike other clients that use hard-coded finite automata machines to parse control sequences, WindTerm uses a configurable finite automata machine that makes it very easy to extend terminal emulators (`terminal/terms/*.term`).

The solution after the garbled is displayed is also very simple, just enter the `echo -e \xf` you mentioned. I will provide a better way to circumvent or solve this problem.

**Comment by git-jiadong** [comment]
kingToolbox, 感谢你的解答，因为我喜欢使用WindTerm看日志，你的高亮出路做得很好。但是因为我在调试看日志的时候避免不了程序日志会出现\xf。我是否能修改(terminal/terms/*.term)来解决我实时看日志乱码的问题呢。如果可以能否提供修改的方法。非常感谢你。我将会在公司推广你的软件，让更多同事来使用，因为这是一款非常棒的终端。非常感谢！

**Comment by kingToolbox** [comment]
Yes, you can. Because the character set switch is caused by `\xe` or `\xf` in the log being parsed as the control sequence `SI` or `SO`. One of the temporary solutions is to prohibit WindTerm from parsing `\xe` and `\xf`. As I mentioned above, WindTerm uses a configurable finite automata machine that makes it very easy to customize terminal emulator, so you can just follow the steps below. 
- According to the `TERM` of your current session, find the corresponding configuration file. For example, if your current `TERM` in the `session property` dialog is `xterm-256color`, then open the `terminal/terms/xterm-256color.term` file. 
- Find and delete the following two lines:
``` json
 		| (?<ShiftLockingOne>                            \\x0e) # SO
		| (?<ShiftLockingZero>                           \\x0f) # SI
```
- Restart WindTerm

You will find that `\xe` is displayed as `SI` and `\xf` is displayed as `SO`, but the character set will no longer be switched.

Of course, this is only a temporary solution, and the normal character set switching feature will also been removed, such as the table can no longer be drawn correctly. But at least it works, and I will fix this problem in WindTerm_2.2.x version. Please stay tuned, thank you.

**Comment by git-jiadong** [comment]
kingToolbox thank you。我找到了临时解决方案。使用awk过滤要输出到终端的\xe，当我打印日志的时候。
```
adb shell tail -F /tmp/my.log | sed "s/\\x0e//g"
```
这样我就能正常的工作看日志了。
期待你的新版本

**Comment by kingToolbox** [comment]
If you use the Linux command, you can also try `strings`:
```
strings -1 /tmp/my.log 
```

**Comment by git-jiadong** [comment]
感谢你的建议

**Comment by kingToolbox** [comment]
Hi, I just released [WindTerm_2.2.0_Prerelease_1](https://github.com/kingToolbox/WindTerm/releases/tag/2.2-prerelease) and added the `Reset Character Set` feature. If you encounter this problem again, you only need to simply execute the `Right click menu - Reset - Reset Character Set` menu item to restore the normal display. You can download and use it. Thank you.

**Comment by git-jiadong** [comment]
Thank, long-expected
